### ISSUE TITLE
Feature request: user configurable controls

### ISSUE THREAD

**Issue by sudomain** [bk138/multivnc#58]
If you'd like to put out an incentive for implementing this feature, you can do so at https://issuehunt.io/r/bk138/multivnc?tab=idle

**Is your feature request about the Desktop Multivnc or the Mobile MultiVNC?**
 - Desktop (wxWidgets) or Mobile (Android): Mobile
 - If Mobile, installed from Play Store or F-Droid: F-Droid

**Is your feature request related to a problem? Please describe.**

There are a few controls I'd like to be able to configure. I'd like to be able to:
1. Make a long press gesture send the right click signal (rather than open the zoom controls)
2. Make a double tap gesture send a double click
3. Make a double tap and hold gesture (i.e. tap, release, tap and hold on the second tap) send a click and hold signal (i.e. click and drag)

These controls are akin to the controls for the VNC Viewer app on the playstore. I prefer these controls to MultiVNC's default controls. In fact, I only keep that closed source app installed for those controls.

**Describe the solution you'd like**

Rather than impose my preference on all users of MultiVNC, it would be great if users could configure  their preferred controls. Maybe a menu to configure key-value pairs of "gesture:signal"

**Describe alternatives you've considered**

**Additional context**

**Comment by bk138** [comment]
You might want to have a look at [AndroidVNC](https://play.google.com/store/apps/details?id=android.androidVNC&hl=en_US), which MultiVNC was forked from back then. It still includes a bunch of other input modes that were deliberately removed from MultiVNC to keep things simple.

**Comment by sudomain** [comment]
I tried AndroidVNC a while back and ended up going with MultiVNC. After trying the various input modes again today, I found that none of them will do what I want (double tap to double click and long press to right click) nor is there a way to configure this.

It turns out that the feature to [remap doubleclick and right click  for AndroidVNC goes back to 2009](https://github.com/antlersoft/android-vnc-viewer/issues/104).

Would a PR for a remapping feature be welcomed? I can't spare the time right now, but maybe in a few months

**Comment by bk138** [comment]
TBH, I dont want to add a lot of configuration options, _but_ I'm open to sensible defaults.

As it stands, I would not want to change anything w.r.t. virtual mouse button mode as this is pretty much self-explaining and there should definitely be no other way in there to inject clicks as per the virtual mouse buttons.

For the buttonless mode, we could work out more sensible defaults, so I'd like to keep this issue open for community feedback.

**Comment by gujjwal00** [comment]
If/when current defaults are changed, we should change the action for two finger swipe/fling from arrow keys to panning. 

Some gestures can cause interleaved scroll/scale events. Panning on scroll will play nicely with scaling. This is how most other apps handle these gestures (ex: Firefox, Google PDF Viewer, Gallery) . It will also simplify our code handling these gestures.

We will then have to decide on how to invoke arrow keys. We could use three finger swipe (which is used for panning right now).

**Comment by sudomain** [comment]
> If/when current defaults are changed, we should change the action for two finger swipe/fling from arrow keys to panning.

Is panning the same/what's caused by scrolling on a hardware mouse wheel? I'm also for removal of arrow keys. My opinion probably shouldn't matter as much as someone's who has contributed code to this project, but I'd be totally ok with removal of arrow keys without replacement. Imo arrow keys should be the responsibility of the user's keyboard app and not the VNC client. There is always the "Key Combo" menu item though

**Comment by gujjwal00** [comment]
> Is panning the same/what's caused by scrolling on a hardware mouse wheel?

Yes.  Panning shifts the visible part of frame (if you are zoomed in). This is currently done by 3-finger swipe. 

>Imo arrow keys should be the responsibility of the user's keyboard app and not the VNC client

Another problem is that if I have to send multiple arrow keys, I can keep the button pressed but with these gestures, it gets annoying really fast. 

Current solutions:
- Change gesture for arrow keys: most probably to 3-finger swipe 
- Replace gesture with 'virtual' arrow buttons
- Remove them without replacement

PS: I can't even use 3-finger swipes because they are intercepted by my device for global gestures (like taking screenshots). I have to move the pointer to force panning.

**Comment by anonimo82** [comment]
I'd like this solution implemented too, especially I'd like some way to scroll from MultiVNC Android App

**Comment by anonimo82** [comment]
Is AVNC OpenGL enabled as MultiVNC?

**Comment by sudomain** [comment]
It appears so?
https://github.com/gujjwal00/avnc/search?q=opengl&type=

**Comment by anonimo82** [comment]
No, tried it. Sooooo slow

**Comment by bk138** [comment]
@anonimo82 AVNC definitely uses OpenGL drawing. It does things a bit differently than MultiVNC but should be no slower.
